app/models/user.rb

```diff
@@ -37,12 +39,30 @@ def remember
     update_attribute(:remember_digest, User.digest(remember_token))
   end
 
-  def authenticated? remember_token
-    return false unless remember_digest
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def authenticated? attribute, token
+    digest = send("#{attribute}_digest")
+    return false unless digest
+    BCrypt::Password.new(digest).is_password?(token)
   end
 
   def forget
     update_attribute :remember_digest, nil
   end
+
+  def downcase_email
+    self.email = email.downcase
```

Comment:
Sao phải dùng `   self.email = email.downcase`  mà ko giữ code cũ `email.downcase!`

Reply:
Ở code cũ thì sẽ downcase email khi tạo mới và cập nhật người dùng. Còn ở trường hợp này chỉ cần downcase email khi tạo mới người dùng. Em thấy trong book nói vậy, nhưng em thử sửa đổi email người dùng chứa kí tự IN HOA, nhưng trong database vẫn set email là chữ thường. Em vẫn đang thắc mắc chỗ đó.

Reply:
2 code chị hỏi đều cho kết quả như nhau đấy

Reply:
Chính vì 2 code có tác dụng giống nhau, mà trong book em thấy giải thích giống như ở trên nên em cũng thắc mắc chỗ đó đó chị.

Reply:
lên đây chị chỉ cho coi :p, tui nhát gõ quá